File: android/src/main/java/com/imagefilters/FilterRegistry.kt
package com.imagefilters

/**
 * Registry for managing available filters
 */
object FilterRegistry {
    
    private val documentFilters = setOf(
        "scan", "blackWhite", "enhance", "perspective", "grayscale", "colorPop"
    )
    
    private val photoFilters = setOf(
        "sepia", "noir", "fade", "chrome", "transfer", "instant",
        "vivid", "dramatic", "warm", "cool", "vintage",
        "clarendon", "gingham", "juno", "lark", "luna", "reyes", "valencia",
        "brooklyn", "earlybird", "hudson", "inkwell", "lofi", "mayfair",
        "nashville", "perpetua", "toaster", "walden", "xpro2"
    )
    
    private val customFilters = mutableSetOf("custom")
    
    /**
     * Get all available filters
     */
    fun getAllFilters(): List<String> {
        return (documentFilters + photoFilters + customFilters).sorted()
    }
    
    /**
     * Get filters by type
     */
    fun getFilters(type: String?): List<String> {
        return when (type?.lowercase()) {
            "document" -> documentFilters.sorted()
            "photo" -> photoFilters.sorted()
            "custom" -> customFilters.sorted()
            else -> getAllFilters()
        }
    }
    
    /**
     * Check if filter is valid
     */
    fun isValid(filter: String): Boolean {
        return documentFilters.contains(filter) ||
                photoFilters.contains(filter) ||
                customFilters.contains(filter)
    }
    
    /**
     * Get filter category
     */
    fun getCategory(filter: String): FilterCategory {
        return when {
            documentFilters.contains(filter) -> FilterCategory.DOCUMENT
            photoFilters.contains(filter) -> FilterCategory.PHOTO
            customFilters.contains(filter) -> FilterCategory.CUSTOM
            else -> FilterCategory.UNKNOWN
        }
    }
    
    /**
     * Register custom filter
     */
    fun registerCustomFilter(name: String) {
        customFilters.add(name)
    }
    
    /**
     * Unregister custom filter
     */
    fun unregisterCustomFilter(name: String) {
        customFilters.remove(name)
    }
}

/**
 * Filter category enum
 */
enum class FilterCategory {
    DOCUMENT,
    PHOTO,
    CUSTOM,
    UNKNOWN
}

/**
 * Filter parameters data class
 */
data class FilterParameters(
    var intensity: Float = 1.0f,
    var customParams: Map<String, Any> = emptyMap()
) {
    fun getFloat(key: String, defaultValue: Float = 1.0f): Float {
        return when (val value = customParams[key]) {
            is Number -> value.toFloat()
            else -> defaultValue
        }
    }
    
    fun getInt(key: String, defaultValue: Int = 0): Int {
        return when (val value = customParams[key]) {
            is Number -> value.toInt()
            else -> defaultValue
        }
    }
    
    fun getBoolean(key: String, defaultValue: Boolean = false): Boolean {
        return when (val value = customParams[key]) {
            is Boolean -> value
            else -> defaultValue
        }
    }
}

File: ios/ImageFilters/FilterRegistry.swift
import Foundation
import Metal
import CoreImage

/// Registry for managing available filters
@available(iOS 18.0, *)
class FilterRegistry {
    
    // MARK: - Singleton
    
    static let shared = FilterRegistry()
    
    // MARK: - Properties
    
    private var documentFilters: Set<String> = [
        "scan", "blackWhite", "enhance", "perspective", "grayscale", "colorPop"
    ]
    
    private var photoFilters: Set<String> = [
        "sepia", "noir", "fade", "chrome", "transfer", "instant",
        "vivid", "dramatic", "warm", "cool", "vintage",
        "clarendon", "gingham", "juno", "lark", "luna", "reyes", "valencia",
        "brooklyn", "earlybird", "hudson", "inkwell", "lofi", "mayfair",
        "nashville", "perpetua", "toaster", "walden", "xpro2"
    ]
    
    private var customFilters: Set<String> = ["custom"]
    
    // MARK: - Public Methods
    
    /// Get all available filters
    func getAllFilters() -> [String] {
        return Array(documentFilters.union(photoFilters).union(customFilters)).sorted()
    }
    
    /// Get filters by type
    func getFilters(type: String?) -> [String] {
        switch type?.lowercased() {
        case "document":
            return Array(documentFilters).sorted()
        case "photo":
            return Array(photoFilters).sorted()
        case "custom":
            return Array(customFilters).sorted()
        default:
            return getAllFilters()
        }
    }
    
    /// Check if filter is valid
    func isValid(filter: String) -> Bool {
        return documentFilters.contains(filter) ||
               photoFilters.contains(filter) ||
               customFilters.contains(filter)
    }
    
    /// Get filter category
    func getCategory(for filter: String) -> FilterCategory {
        if documentFilters.contains(filter) {
            return .document
        } else if photoFilters.contains(filter) {
            return .photo
        } else if customFilters.contains(filter) {
            return .custom
        }
        return .unknown
    }
    
    /// Register custom filter
    func registerCustomFilter(name: String) {
        customFilters.insert(name)
    }
    
    /// Unregister custom filter
    func unregisterCustomFilter(name: String) {
        customFilters.remove(name)
    }
}

// MARK: - Filter Category

enum FilterCategory {
    case document
    case photo
    case custom
    case unknown
}

// MARK: - Filter Parameters

struct FilterParameters {
    var intensity: Float = 1.0
    var customParams: [String: Any] = [:]
    
    func getFloat(_ key: String, defaultValue: Float = 1.0) -> Float {
        if let value = customParams[key] as? Float {
            return value
        } else if let value = customParams[key] as? Double {
            return Float(value)
        } else if let value = customParams[key] as? Int {
            return Float(value)
        }
        return defaultValue
    }
    
    func getInt(_ key: String, defaultValue: Int = 0) -> Int {
        if let value = customParams[key] as? Int {
            return value
        } else if let value = customParams[key] as? Double {
            return Int(value)
        } else if let value = customParams[key] as? Float {
            return Int(value)
        }
        return defaultValue
    }
    
    func getBool(_ key: String, defaultValue: Bool = false) -> Bool {
        return customParams[key] as? Bool ?? defaultValue
    }
}

